src/System.Net.HttpListener/src/System/Net/WebSockets/WebSocketProtocolComponent.cs

```diff
@@ -0,0 +1,709 @@
+                "'webSocket.SessionHandle' MUST NOT be NULL or INVALID.");
+            Debug.Assert(actionContext != IntPtr.Zero, "'actionContext' MUST NOT be IntPtr.Zero.");
+            Debug.Assert(bytesTransferred >= 0, "'bytesTransferred' MUST NOT be negative.");
+
+            if (webSocket.SessionHandle.IsClosed)
+            {
+                return;
+            }
+
+            try
+            {
+                WebSocketCompleteAction_Raw(webSocket.SessionHandle, actionContext, (uint)bytesTransferred);
+            }
+            catch (ObjectDisposedException)
+            {
+            }
+        }
+
+        private static void DrainActionQueue(SafeHandle webSocketHandle, ActionQueue actionQueue)
+        {
+            Debug.Assert(webSocketHandle != null && !webSocketHandle.IsInvalid,
+                "'webSocketHandle' MUST NOT be NULL or INVALID.");
+
+            IntPtr actionContext;
+            IntPtr dummy;
+            Action action;
+            BufferType bufferType;
+
+            while (true)
+            {
+                Buffer[] dataBuffers = new Buffer[1];
+                uint dataBufferCount = 1;
+                int errorCode = WebSocketGetAction_Raw(webSocketHandle,
+                    actionQueue,
+                    dataBuffers,
+                    ref dataBufferCount,
+                    out action,
+                    out bufferType,
+                    out dummy,
+                    out actionContext);
+
+                if (!Succeeded(errorCode))
+                {
+                    Debug.Assert(errorCode == 0, "'errorCode' MUST be 0.");
+                    return;
+                }
+
+                if (action == Action.NoAction)
+                {
+                    return;
+                }
+
+                WebSocketCompleteAction_Raw(webSocketHandle, actionContext, 0);
+            }
+        }
+
+        private static void MarshalAndVerifyHttpHeader(IntPtr httpHeaderPtr,
+            ref HttpHeader httpHeader)
+        {
+            Debug.Assert(httpHeaderPtr != IntPtr.Zero, "'currentHttpHeaderPtr' MUST NOT be IntPtr.Zero.");
+
+            IntPtr httpHeaderNamePtr = Marshal.ReadIntPtr(httpHeaderPtr);
+            IntPtr lengthPtr = IntPtr.Add(httpHeaderPtr, IntPtr.Size);
+            int length = Marshal.ReadInt32(lengthPtr);
+            Debug.Assert(length >= 0, "'length' MUST NOT be negative.");
+
+            if (httpHeaderNamePtr != IntPtr.Zero)
+            {
+                httpHeader.Name = Marshal.PtrToStringAnsi(httpHeaderNamePtr, length);
+            }
+
+            if ((httpHeader.Name == null && length != 0) ||
+                (httpHeader.Name != null && length != httpHeader.Name.Length))
+            {
+                Debug.Assert(false, "The length of 'httpHeader.Name' MUST MATCH 'length'.");
+                throw new AccessViolationException();
+            }
+
+            // structure of HttpHeader:
```

Comment:
This is a bad way to emulate a struct: we risk missing padding and corrupting memory on other architectures.